learning-development/r.qmd

```diff
@@ -248,9 +248,158 @@ R can do a lot of the things that are possible in SQL. The reference table below
+
+### Executing SQL scripts from within R
+
+This is primarily useful if you already have existing SQl scripts that you want to build into an R controlled pipeline. For using R to execute a SQL script you'll need the SQL script/s to be in your R Project and to make a connection via R to the database.
+
+```
+# Library calls ====
+
+library(odbc)
+library(DBI)
+
+# DB connection ====
+
+con <- DBI::dbConnect(odbc::odbc(),
+                      Driver = "ODBC Driver 17 for SQL Server",
+                      Server = "server_name",
+                      Database = "database_name",
+                      UID = "",
+                      PWD = "",
+                      Trusted_Connection = "Yes"
+)
+
+# Function to read in sql scripts ====
+
+getSQL <- function(filepath){
+  con = file(filepath, "r")
+  sql.string <- ""
+  
+  while (TRUE){
+    line <- readLines(con, n = 1)
+    
+    if ( length(line) == 0 ){
+      break
+    }
+    
+    line <- gsub("\\t", " ", line)
+    
+    if(grepl("--",line) == TRUE){
+      line <- paste(sub("--","/*",line),"*/")
+    }
+    
+    sql.string <- paste(sql.string, line)
+  }
+  
+  close(con)
+  return(sql.string)
+}
+
+# Execute SQL query and pull into R ====
+
+my_table <- dbGetQuery(con, getSQL("my_query_script.sql"))
+
+```
+
+If you're struggling to get your SQL scripts to execute from R, try adding the following two lines to the top of your SQL script to force the formatting required for it to work:
+
+
+```
+SET ANSI_PADDING OFF
+SET NOCOUNT ON;
+```
+
+---
+
+### Using dbplyr to create database queries
+
+If you're familiar with dplyr then using the dbplyr package to interact with a SQL database is a good way to go. It allows you to use common dplyr functions such as select(), filter(), mutate(), group_by(), summarise() and arrange() with a SQL database.
+
+As with sending a SQL script, you'll need to create the connection to the database first:
+
+```
+# Library calls ====
+
+library(odbc)
+library(DBI)
+
+# DB connection ====
+
+con <- DBI::dbConnect(odbc::odbc(),
+                      Driver = "ODBC Driver 17 for SQL Server",
+                      Server = "server_name",
+                      Database = "database_name",
+                      UID = "",
+                      PWD = "",
+                      Trusted_Connection = "Yes"
+)
+```
+
+Once you've established a connection, you can then create piped queries as you would with dplyr:
+
+```
+query_result <- collect(
+  tbl(con, Id(schema='dbo', table='table_name')) %>%
+  select(example_id_column, example_filter_column, example_indicator_column) %>%
+  filter(example_filter_column=='Filter entry 1', example_indicator_column <= 120)
+)
```

Comment:
I've always written this with collect separately like the `x <- tbl(...) %>% select(...) %>% collect()`, so you do all the things and then collect the results at the end. I'm happy either way, just curious if it makes any difference or if there's a preferred way we could be consistent with?

Reply:
Both syntax styles work identically. Either way, it collates the commands and then collects the data. Each of the different styles probably make different sense to different people. I just put everything inside the collect function as I thought that fitted better conceptually with the idea that it's evaluating all the rest of the commands that preceded it in one go, whereas the pipe notation kind of makes it look like each command is fully run sequentially.